CDAP-15130: Implement Salesforce Batch Multi Source plugin

aielchiieva opened the pull request:
1. Factored out common logic for Salesforce Batch Source and Multi Source plugins into base package.
2. Added Salesforce Batch Multi Source classes into multi package.
3. Added integration tests.
4. Fixed parser to cover cases when SObject contains field with the same name as SObject name and when SObject contains field named Category which was treated as keyword in parser.

jayfjin commented on docs/SalesforceMultiObjects-batchsource.md:
```diff
@@ -0,0 +1,121 @@
+    
+To read data from these two SObjects, both of them must be indicated in the `White List` configuration property.
+
+The output of the the source will be the following records:
```
I think it will be helpful to also show here how the schemas are merged into the record (with the UUID of each).

jayfjin commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseBatchSource.java:
```diff
@@ -0,0 +1,148 @@
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      switch (logicalType) {
+        case DATE:
```
I've seen this switch case structure around a few times now in various code reviews. Does this refactor capture all usages of it? If not we should extract this to a util/helper class

jayfjin commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseSourceConfig.java:
```diff
@@ -0,0 +1,145 @@
+   *
+   * @return SOQL generated based on sObject metadata and given filters
+   */
+
```
nit: extra newline

jayfjin commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/multi/SalesforceMultiInputFormatProvider.java:
```diff
@@ -0,0 +1,95 @@
+    fields.add(Schema.Field.of(sObjectNameField, Schema.of(Schema.Type.STRING)));
+    return Schema.Field.of(field.getName(), Schema.recordOf(fieldSchema.getRecordName(), fields));
+  }
+
```
nit: extra newline

jayfjin commented on src/test/java/io/cdap/plugin/salesforce/etl/SalesforceBatchMultiSourceETLTest.java:
```diff
@@ -0,0 +1,137 @@
+
+    String sObjectNameField = SalesforceMultiSourceConfig.SOBJECT_NAME_FIELD_DEFAULT;
+
+    Set<String> expectedResults = Sets.newHashSet(
```
Instead of converting the record into a string, can we build what the expected records should look like for each, and then compare the records?

albertshau replied:
If the values are just strings, you be able to build the records like Jay mentioned and Assert.assertEquals(expectedRecords, actualRecords).

It's only if the records contain bytes, arrays, and other things that don't compare nicely in Java that you need to explicitly compare field values.

aielchiieva replied:
Done.

jayfjin commented on src/main/java/io/cdap/plugin/salesforce/SalesforceSchemaUtil.java:
```diff
@@ -79,7 +81,38 @@ public static Schema getSchema(AuthenticatorCredentials credentials, SObjectDesc
+  }
+
+  /**
+   * Generates CDAP schema for multiple SObjects, where each field of type record represents one SObjects.
```
Is this only used to generate CDAP schema in the multiple SObject case? If so imo the function name should reflect that

albertshau commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseBatchSource.java:
```diff
@@ -0,0 +1,148 @@
+      case DOUBLE:
+        return Double.parseDouble(value);
+      case BYTES:
+        return Byte.parseByte(value);
```
this is wrong, it is parsing a single byte whereas 'bytes' is a byte array. Does salesforce have a bytes type?

albertshau commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseBatchSource.java:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @param <T> input data key value
+ */
+public abstract class SalesforceBaseBatchSource<T> extends BatchSource<T, Map<String, String>, StructuredRecord> {
```
Instead of an abstract base plugin class, this logic should be pulled out into a class that transforms a single line from the bulk API response to a StructuredRecord.

albertshau commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseSourceConfig.java:
```diff
@@ -0,0 +1,145 @@
+          .filter(name -> schema.getField(name) != null)
+          .collect(Collectors.toList());
+
+        if (fieldNames.isEmpty()) {
```
Is there validation done somewhere checking if the schema has fields that are not present in the SObject?

albertshau commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/base/SalesforceBaseSourceConfig.java:
```diff
@@ -0,0 +1,145 @@
+      LOG.debug("Generated SObject query: '{}'", sObjectQuery);
+      return sObjectQuery;
+    } catch (ConnectionException e) {
+      throw new IllegalStateException(
```
doesn't look like this is adding much value, this method should just throw the ConnectionException.

albertshau commented on src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/multi/SalesforceMultiInputFormat.java:
```diff
@@ -0,0 +1,82 @@
+  @Override
+  public List<InputSplit> getSplits(JobContext context) {
+    Configuration configuration = context.getConfiguration();
+    List<String> queries = GSON.fromJson(configuration.get(SalesforceSourceConstants.CONFIG_QUERIES), QUERIES_TYPE);
```
Instead of having separate Multi classes for the input format, record reader, split, and inputformat provider, it is much simpler to just add the query to the SalesforceSplit and enhance the existing inputformat to support multiple queries. The multi source would just configure it with multiple queries while the single source would configure it with a single query.

aielchiieva replied:
Done, though I felt code was less coupled with previous design.

aielchiieva commented:
@albertshau 

> On a separate note, I looked over the existing SalesforceRecordReader and noticed it reads the entire Bulk API response into a single string, then splits and parses.
> 
> This is not correct, the record reader should be reading the response in a streaming fashion, with each call to nextKeyValue reading the next line of input from the input stream, and close closing the input stream. Please open a jira and fix it in a separate PR.

It works not quite as you have described. `SalesforceInputFormat#getSplits` submits Bulk API query and receives batchId array. For each batchId new input split is created. `SalesforceRecordReader` queries Salesforce using batchId from input split. As result it receives list of input streams for one batchId. Since all input streams are the parts of one CSV file they should be read into a single string to be parsed properly using CSV reader.

aielchiieva commented:
@albertshau / @jayfjin addressed code review comment, please review.

aonischuk commented on docs/SalesforceMultiObjects-batchsource.md:
```diff
@@ -0,0 +1,121 @@
+Offset units set to `hours`. For example, if duration is '6' (6 hours) and the offset is '1' (1 hour) and the pipeline 
+runs at 9am, it will read data updated from 2am - 8am. Ignored if `datetimeFilter` is provided.
+
+**Error Handling:** Strategy used to handle erroneous records. Acceptable values are Skip on error,
```
This function was requested by Nitin Motgi 
https://github.com/data-integrations/salesforce/pull/4#discussion_r268378379

The error happens if the csv row is invalid. For example not enough values or too much values or casting to type from schema is not valid (for example expected in found string)

albertshau replied:
but doesn't the schema come from salesforce? Does it somehow allow its data to not match its schema?

aonischuk replied:
Yes schema comes from Salesforce. I don't think it should ever happen in normal flow.

We just had some handling for unexpected values in CSV and Nitin asked to implement error handling like that. 

@nitinmotgi do you think we can remove errorHandling due to reasons Albert mentions or you have some other cases in mind?

albertshau replied:
Let's open a jira for this so that it doesn't block this PR in the meantime.

aielchiieva replied:
Created Jira - https://issues.cask.co/browse/CDAP-15325

albertshau commented on docs/SalesforceMultiObjects-batchsource.md:
```diff
@@ -0,0 +1,121 @@
+
+**Black List**: List of SObjects NOT to replicate from Salesforce. By default NONE of SObjects will be black listed.
+
+**Datetime Filter:** SObject query datetime filter is applied to system field `LastModifiedDate` to allow incremental 
```
OK let's open a jira and fix separately.

After reading the doc again I understand better, this is only used when an SObject is given and not for SOQL. I think the confusing part is that users can either specify a DatetimeFilter (which is really just specifying the start time), or they can specify a duration and offset.

I think the confusion can be removed if instead of these three, we have a 'start' and 'end' property. If 'start' is not specified, it reads from the start of time. If 'end' is not specified, it reads up until the logical start time of the pipeline (context.getLogicalStartTime()). This way, 'start' replaces the current DatetimeFilter, and it is probably easier to understand than duration and offset.

aielchiieva replied:
Created Jira for the fix: https://issues.cask.co/browse/CDAP-15320

albertshau commented:
At https://github.com/data-integrations/salesforce/blob/master/src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceRecordReader.java#L78, it is reading the entire contents of the response into memory, then parsing.

https://github.com/data-integrations/salesforce/blob/master/src/main/java/io/cdap/plugin/salesforce/SalesforceBulkUtil.java#L113 is a bad method and should not exist. The InputStream from https://github.com/data-integrations/salesforce/blob/master/src/main/java/io/cdap/plugin/salesforce/SalesforceBulkUtil.java#L127 should be used to read the next line whenever the record reader gets the next key value, and closed when the record reader is closed.

albertshau commented:
just one comment regarding one of Jay's comments in the test

aielchiieva commented:
@albertshau updated integration test. If all changes look good, I will squash the commits and resolve the conflicts.

aielchiieva commented:
@albertshau 
> At https://github.com/data-integrations/salesforce/blob/master/src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceRecordReader.java#L78, it is reading the entire contents of the response into memory, then parsing later in the record reader.
> 
> https://github.com/data-integrations/salesforce/blob/master/src/main/java/io/cdap/plugin/salesforce/SalesforceBulkUtil.java#L113 is a bad method and should be deleted. The InputStream from https://github.com/data-integrations/salesforce/blob/master/src/main/java/io/cdap/plugin/salesforce/SalesforceBulkUtil.java#L127 should be used to read the next line whenever the record reader gets the next key value, and closed when the record reader is closed.

I have created Jira for further discussion and fix - https://issues.cask.co/browse/CDAP-15319.

aielchiieva commented on docs/SalesforceMultiObjects-batchsource.md:
```diff
@@ -0,0 +1,121 @@
+`LAST_MONTH` ...
+
+**Duration:** SObject query duration filter is applied to system field `LastModifiedDate` to allow range query. 
+Duration units set to `hours`. For example, if duration is '6' (6 hours) and the pipeline runs at 9am, it will read data 
```
Created Jira for the fix: https://issues.cask.co/browse/CDAP-15320

aielchiieva commented:
@albertshau / @jayfjin addressed remaining code review comments, for each mentioned enhancement has created Jiras (only one is under discussion about error handling, waiting for the comment from @nitinmotgi). Please let me know if there are any other changes required before I squash and resolve the conflicts.

aielchiieva commented:
@albertshau squashed commits and resolved the conflicts. PR is ready to be merged.
